zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,970 @@
+                config.msg_hash[i],
+                offset,
+                || Ok(F::from(*byte as u64)),
+            )?;
+        }
+
+        Ok((
+            AssignedSignatureVerify {
+                address: address_assigned,
+                msg_hash_rlc: msg_hash_rlc_assigned,
+            },
+            KeccakAux {
+                input: pk_bytes_be,
+                output: pk_hash.try_into().expect("vec to array of size 32"),
+            },
+        ))
+    }
+
+    pub fn assign(
+        &self,
+        config: &SignVerifyConfig<F>,
+        layouter: &mut impl Layouter<F>,
+        randomness: F,
+        signatures: &[SignData],
+    ) -> Result<Vec<AssignedSignatureVerify<F>>, Error> {
+        if signatures.len() > MAX_VERIF {
+            error!(
+                "signatures.len() = {} > MAX_VERIF = {}",
+                signatures.len(),
+                MAX_VERIF
+            );
+            return Err(Error::Synthesis);
+        }
+        let main_gate = MainGate::new(config.main_gate_config.clone());
+        // TODO: Figure out the best value for RangeChip base_bit_len, when we want to
+        // range on 8 bits.
+        let range_chip = RangeChip::new(config.range_config.clone(), 8);
+        let mut ecc_chip = GeneralEccChip::<Secp256k1Affine, F, NUMBER_OF_LIMBS, BIT_LEN_LIMB>::new(
+            config.ecc_chip_config(),
+        );
+        let scalar_chip = ecc_chip.scalar_field_chip();
+
+        layouter.assign_region(
+            || "ecc chip aux",
+            |mut region| self.assign_aux(&mut region, &mut ecc_chip),
+        )?;
+
+        let ecdsa_chip = EcdsaChip::new(ecc_chip.clone());
+        let address_is_zero_chip = IsZeroChip::construct(config.address_is_zero.clone());
+
+        let mut assigned_ecdsas = Vec::new();
+        let mut keccak_auxs = Vec::new();
+
+        let chips = ChipsRef {
+            main_gate: &main_gate,
+            range_chip: &range_chip,
+            ecc_chip: &ecc_chip,
+            scalar_chip: &scalar_chip,
+            ecdsa_chip: &ecdsa_chip,
+        };
+
+        layouter.assign_region(
+            || "ecdsa chip verification",
+            |mut region| {
+                assigned_ecdsas.clear();
+                keccak_auxs.clear();
+                let offset = &mut 0;
+                let mut ctx = RegionCtx::new(&mut region, offset);
+                for i in 0..MAX_VERIF {
+                    let signature = if i < signatures.len() {
+                        signatures[i].clone()
+                    } else {
+                        // padding (enabled when address == 0)
+                        SignData::default()
+                    };
+                    let assigned_ecdsa = self.assign_ecdsa(&mut ctx, &chips, &signature)?;
+                    assigned_ecdsas.push(assigned_ecdsa);
+                }
+                Ok(())
+            },
+        )?;
+
+        let mut assigned_sig_verifs = Vec::new();
```

Comment:
Why the `.clear()` here?

Reply:
Ah, very good question, because this is not obvious.  So halo2 will call the closure passed to `assign_region` twice!  If I remember correctly, this is related to how the layouter works, first it runs the closure to figure out the area required for the region, then it fins a place for each region, and then it runs the closure again to assign the cell values at their final destination.  On the first run, no witness data is used.

In `assigned_sig_verifs` I want the final assigned cells, so I added this `.clear()` so that I always get the result of the last pass (the good one).

Reply:
Does the circuit fail to work without `.clear()`?
Do we worry about `assigned_sig_verifs` getting contaminated in the first call? But `assigned_sig_verifs` should be refreshed between 2 `assign` calls, so it should be fine to remove the `.clear()`.

Reply:
> Does the circuit fail to work without `.clear()`? 

Yes it fails without the `.clear()`.  The reason is because in the `assign_region` I'm pushing assigned cells to the vector, and later on I'm accessing the assigned cells by index.  Without the `.clear()`, I'd be accessing invalid vales (the result of the first pass).

> Do we worry about `assigned_sig_verifs` getting contaminated in the first call? But `assigned_sig_verifs` should be refreshed between 2 `assign` calls, so it should be fine to remove the `.clear()`.

Imagine the circuit is set up for 2 verifications.  Without the `.clear()`, `assigned_sig_verifs` will contain 4 elements, where only the last 2 are the assigned cells I want.